### PR TITLE
Invalid warning when compiling frontend with gcc 13

### PR DESCRIPTION
**Pull request by stonea**
Recently on my desktop machine I updated to gcc 13.1.1, ever since then I get these error when compiling Chapel:

```
[  0%] Building CXX object frontend/lib/CMakeFiles/ChplFrontend-obj.dir/util/clang-integration.cpp.o
/home/stonea/chapel/frontend/lib/util/clang-integration.cpp: In function ‘chpl::owned<chpl::TemporaryFileResult>& chpl::util::createClangPrecompiledHeader(chpl::Context*, chpl::ID)’:
/home/stonea/chapel/frontend/lib/util/clang-integration.cpp:252:37: error: possibly dangling reference to a temporary [-Werror=dangling-reference]
  252 |     const std::vector<std::string>& cc1args =
      |                                     ^~~~~~~
/home/stonea/chapel/frontend/lib/util/clang-integration.cpp:253:24: note: the temporary was destroyed at the end of the full expression ‘chpl::util::getCC1Arguments(context, std::vector<std::__cxx11::basic_string<char> >(clFlags), 0)’
  253 |         getCC1Arguments(context, clFlags, /* forGpuCodegen */ false);
      |         ~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/stonea/chapel/frontend/lib/util/clang-integration.cpp: In function ‘const bool& chpl::util::precompiledHeaderContainsNameQuery(chpl::Context*, const chpl::TemporaryFileResult*, chpl::UniqueString)’:
/home/stonea/chapel/frontend/lib/util/clang-integration.cpp:330:37: error: possibly dangling reference to a temporary [-Werror=dangling-reference]
  330 |     const std::vector<std::string>& cc1args =
      |                                     ^~~~~~~
/home/stonea/chapel/frontend/lib/util/clang-integration.cpp:331:22: note: the temporary was destroyed at the end of the full expression ‘chpl::util::getCC1Arguments(context, std::vector<std::__cxx11::basic_string<char> >(clFlags), 0)’
  331 |       getCC1Arguments(context, clFlags, /* forGpuCodegen */ false);
      |       ~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1plus: all warnings being treated as errors
make[6]: *** [frontend/lib/CMakeFiles/ChplFrontend-obj.dir/build.make:2596: frontend/lib/CMakeFiles/ChplFrontend-obj.dir/util/clang-integration.cpp.o] Error 1
make[5]: *** [CMakeFiles/Makefile2:819: frontend/lib/CMakeFiles/ChplFrontend-obj.dir/all] Error 2
make[4]: *** [CMakeFiles/Makefile2:774: compiler/CMakeFiles/chpl.dir/rule] Error 2
make[3]: *** [Makefile:169: chpl] Error 2
make[2]: *** [Makefile:197: /home/stonea/chapel/bin/linux64-x86_64/chpl] Error 2
make[1]: *** [Makefile:82: compiler] Error 2
make: *** [Makefile:59: comprt] Error 2
```

What's happening is for whatever reason this function returns a reference to a vector:

```
const std::vector<std::string>& getCC1Arguments(...)
```

And when we call it we do something like this:

```
const std::vector<std::string>& cc1args = getCC1Arguments(...)
```

~~[Looking internally into `getCC1Arguments`](https://github.com/chapel-lang/chapel/blob/827df4f0d2281c44a4bffe2cb866a7503d9a92cf/frontend/lib/util/clang-integration.cpp#L198) I think? it's returning a variable on its stack (`result`), so it's true that the thing being referenced to is going to get destroyed after finishing the call.  I can get the issue to go away by just having the call site not assign to a reference.  Maybe I'm misunderstanding something, but I think with the move construction and/or copy elision there shouldn't be any unnecessary copies happening anyway.~~

**Edit**: In this case `getCC1Arguments` will indeed return a reference to heap allocated data so the warning is spurious.

TODO:
- [x] Normal paratests

### CODE REVIEW

**Comment by mppf** [comment]
@stonea - I don't like the approach of making a copy of the vector just to silence the noisy GCC warning.

Please see PR #22211 which faced this warning earlier. It is a spurious warning. In that PR, I massaged the code to make the warning go away. My current opinion is that we should un-do that change and instead move to adding `-Wno-error=dangling-reference` or something to disable the warning or make it not an error. But, it would also be OK with me to make code-massaging changes like the earlier PR. Adding a `std::move` at the call site might help, but you might need to do that in a wrapper function, I do not know off-hand.

**Comment by mppf** [comment]
> I think? it's returning a variable on its stack (result), so it's true that the thing being referenced to is going to get destroyed after finishing the call.

It is not, it is returning a reference to something allocated on the heap, and that is part of what `return QUERY_END(result)`.

**Comment by stonea** [comment]
> It is not, it is returning a reference to something allocated on the heap, and that is part of what return QUERY_END(result) does.

Ah ok, yeah I was assuming `QUERY_END` was just populating result and returning a reference to it. So yeah, if its returning something into heap allocated memory we would get an extra copy if I applied my PR as I have it.

> My current opinion is that we should un-do that change and instead move to adding `-Wno-error=dangling-reference` or something to disable the warning or make it not an error

My kneejerk reaction is against silencing a warning, but from the review in your linked PR it sounds like that was your initial thoughts as well but have since changed your mind as it looks like this issue might keep reapparing if we don't suppress it. 

Anyway, my main concern is getting things unblocked so that I can compile on my desktop, but if you or @DanilaFe can help me implement a more "proper" fix I'd be fine going for that as well.

I'm going to guess that means adding something to `make/compiler/Makefile.gnu`?

Maybe checking for a specific GCC version add adding it to `SQUASH_WARN_GEN_CFLAGS` like we do here: https://github.com/chapel-lang/chapel/blob/main/make/compiler/Makefile.gnu#L188

Does that sound right?

Given that this is a suppressing a warning, I should probably bring it up to the whole team in case there's any objections.  Would you be fine if I deferred to you (or Daniel) for any explanations about why this warning is showing up and why we feel ok with suppressing it?

**Comment by mppf** [comment]
> > My current opinion is that we should un-do that change and instead move to adding `-Wno-error=dangling-reference` or something to disable the warning or make it not an error
> 
> My kneejerk reaction is against silencing a warning, but from the review in your linked PR it sounds like that was your initial thoughts as well but have since changed your mind as it looks like this issue might keep reappearing if we don't suppress it.

Yes, exactly.

> 
> Anyway, my main concern is getting things unblocked so that I can compile on my desktop, but if you or @DanilaFe can help me implement a more "proper" fix (i.e. `-Wno-error=dangling-reference`) I'd be fine going for that as well.
> 
> I'm going to guess that means adding something to `make/compiler/Makefile.gnu`?
> 
> Maybe checking for a specific GCC version add adding it to `SQUASH_WARN_GEN_CFLAGS` like we do here: https://github.com/chapel-lang/chapel/blob/main/make/compiler/Makefile.gnu#L188
> 
> Does that sound right?

You would need to add to `WARN_CXXFLAGS` to impact the compiler build. I'd follow-along from
https://github.com/chapel-lang/chapel/blob/5777125fb853e1e6b2f52613c3ae3887663c6961/make/compiler/Makefile.gnu#L328-L336

We can either make the warning not be an error or just supress the warning altogether.

> Given that this is a suppressing a warning, I should probably bring it up to the whole team in case there's any objections.

I don't think that's strictly necessary but it's fine with me if you want to.

> Would you be fine if I deferred to you (or Daniel) for any explanations about why this warning is showing up and why we feel ok with suppressing it?

Yes, but I will mostly describe it as a GCC warning with false positives and point to https://gcc.gnu.org/bugzilla/show_bug.cgi?id=108165 . We have been seeing a lot of GCC warnings that are too noisy for us; you might notice that Makefile.gnu has quite a few workarounds for these. It's clear that the GCC developers don't expect people to use `-Wall -Werror`.

**Comment by stonea** [comment]
@mppf, I've updated the PR to add `-Wno-dangling-reference` when using gcc 13.

After seeing how much of a precedent we already have for silencing spurious gcc warnings, I feel less concerned about merging it. Would you mind taking a quick second look, and if it looks good I'll go ahead and merge?

File: make/compiler/Makefile.gnu [hunk 1]
```diff
@@ -335,6 +335,16 @@ WARN_CXXFLAGS += -Wno-maybe-uninitialized
 endif
 endif
 
+#
+# Avoid a spurious warning in gcc 13, about a "possibly dangling reference"
+# when a function returns a reference (even though it's something gauranteed to
```
Comment (mppf) [inline make/compiler/Makefile.gnu:340]: gauranteed -> guaranteed

**Review by mppf — APPROVED** [review]
Looks good. 

It would be super helpful if you can update the existing comment in frontend/lib/resolution/resolution-queries.cpp :

```
// this wrapper around resolveFunctionByPoisQuery helps to avoid
// 'possibly dangling reference to a temporary' warnings from GCC 13.
```

Could you just update the comment with a new line

```
// TODO: remove this workaround now that the build uses
// -Wno-dangling-reference
```

Thanks

**Pull request merged**